### PR TITLE
Complex shell quote escaping fails

### PR DESCRIPTION
**Pull request by bricoletc**
I found a case, described in #39, which failed on lsf cluster.
I propose here a solution which made it run successfully.

## Added

- Support for complex quote-expansion via `shlex` [#39]
- Function docstrings in `lsf_config.py` for documentation

Some details of the changes:

#39 issue:
 * The split argument string needs to be rejoined using `shlex`
 * Two unit tests for `lsf_config.py` to ensure quotes are shell-escaped
 * One modified unit test for `lsf_submit.py` to ensure quotes are
    shell-escaped
 * Rewrote `__init__` of `Config` object, to directly concatenate
    parameter lists. Makes downstream argument string production simpler to
    read.
 * Added function docstrings in `lsf_config.py` to make understanding easier

Docs/gitignore:
 * Add vim and venv to gitignore
 * Add example of complex quote escaping in `lsf.yaml` in the readme

The tests  pass- I've not tested this exact version of the code on the cluster, so maybe should do that too.
Cheers!

### CODE REVIEW

**Review by leoisl — APPROVED** [review]
All looks fine to me!

File: README.md [hunk 1]
```diff
@@ -303,7 +303,7 @@ All settings are given with the `rule` name as the key, and the additional clust
 settings as a string ([scalar][yaml-collections]) or list
 ([sequence][yaml-collections]).
 
-#### Examples
+#### Example `lsf.yaml` configuration
```
Comment (mbhall88) [inline README.md:306]:
```suggestion
#### Examples
```

File: README.md [hunk 2]
```diff
@@ -355,6 +355,19 @@ foo: "-P gpu -gpu 'gpu resources'"
 
 The above is also a valid form of the previous example but **not recommended**.
 
+#### Quote-escaping in `lsf.yaml`
```
Comment (mbhall88) [inline README.md:358]:
```suggestion
#### Quote-escaping
```

File: README.md [hunk 3]
```diff
@@ -355,6 +355,19 @@ foo: "-P gpu -gpu 'gpu resources'"
...
+
+Some lsf commands require multiple levels of quote-escaping.
```
Comment (mbhall88) [inline README.md:360]:
```suggestion
Some LSF commands require multiple levels of quote-escaping.
```

File: {{cookiecutter.profile_name}}/lsf_config.py [hunk 4]
```diff
@@ -23,6 +24,11 @@ def get(self, key: str, default: Any = None) -> Any:
 
     @staticmethod
     def args_to_dict(args: str) -> Dict[str, str]:
+        """
+        Converts a string into a dictionary where key/value pairs are consecutive
```
Comment (mbhall88) [inline {{cookiecutter.profile_name}}/lsf_config.py:28]:
```suggestion
        """Converts a string into a dictionary where key/value pairs are consecutive
```

**Review by mbhall88 — CHANGES_REQUESTED** [review]
Great catch @bricoletc! I didn't know shlex had a join function.

Just some tiny doc updates

**Comment by mbhall88** [comment]
Ah actually, I've just realised why the CI is failing. `shlex.join` was only added in python 3.8 (and also black formatting is failing). I would suggest using python 3.5 in your local venv for this project.

**Comment by bricoletc** [comment]
Ok! I found a workaround using python3.5-compatible `shlex` invocations.

For failing the black formatting, I see what you're doing; snakemake supports python>=3.5, but black is for python3.6+ only, so you dont install black using `dev-requirements.txt` else that would fail python3.5 CI. I added a note in `CONTRIBUTING.md` saying you need to install latest black- that means CI will pass.

**Review by mbhall88 — APPROVED** [review]
Thanks a lot Brice

**Pull request merged**